### DISCUSSION TITLE
[과목 2] 제 3장 관리구문

### DISCUSSION THREAD

**Discussion by beolggoolosori** [flataex/sqlp-study#20] [Custom 연습문제] [answered]
문제 1.  

상품 테이블(PRODUCT)에는 상품 ID(id), 상품명(name), 가격(price)이 포함되어 있습니다. 상품 가격이 변경되면, 변경된 정보를 변경된 가격정보 테이블(UPDATE_PRODUCT)에 추가합니다. `UPDATE_PRODUCT`테이블은 상품 ID(update_id), 상품명(update_name), 변경된 가격(update_price)이 포함되어 있습니다. 변경된 가격 정보를 `PRODUCT`의 해당 상품 ID에 업데이트하고, 상품이 존재하지 않는 경우 새로운 상품을 추가하는 동기화 작업을 하려고 합니다.  상품 가격 변경 정보를 처리하는 MERGE 구문을 작성하세요. 
> JHKIM.PRODUCT 및 JHKIM.UPDATE_PRODUCT 정보를 참고해도 됩니다...

---

문제 2. 다음 보기 중 트랜잭션(Transaction) 관련 설명으로 올바르지 않은 것은?
1. 트랜잭션은 데이터베이스에서 하나 이상의 작업을 그룹화하여 원자성(Atomicity), 일관성(Consistency), 격리성(Isolation), 지속성(Durability)을 보장한다.
2. 원자성(Atomicity)은 트랜잭션의 작업들이 모두 성공적으로 완료되거나, 아예 실행되지 않은 상태로 되돌려지는 것을 보장한다.
3. 격리성(Isolation)은 트랜잭션의 작업들이 다른 트랜잭션의 작업에 영향을 주지 않도록 보장한다.
4. 지속성(Durability)은 트랜잭션의 작업이 성공적으로 완료된 후, 해당 작업의 결과가 영구적으로 소실될 수 있다.

---

문제 3. 아래의 SQL 블록 실행 후, 각 계좌의 잔액은 어떻게 될까?
```sql
CREATE TABLE account (
    id NUMBER PRIMARY KEY,
    balance NUMBER
);

INSERT INTO account (id, balance) VALUES (1, 1000);
INSERT INTO account (id, balance) VALUES (2, 2000);

DECLARE
    from_account_id NUMBER := 1;
    to_account_id NUMBER := 2;
    transfer_amount NUMBER := 500;
BEGIN
    UPDATE account SET balance = balance - transfer_amount WHERE id = from_account_id;
    -- 에러발생!!
    UPDATE account SET balance = balance + transfer_amount WHERE id = to_account_id;
    ROLLBACK;
END;
```

1. 계좌 1의 잔액은 500, 계좌 2의 잔액은 2500
2. 계좌 1의 잔액은 1000, 계좌 2의 잔액은 2000
3. 계좌 1의 잔액은 0, 계좌 2의 잔액은 3000
4. 계좌 1의 잔액은 1500, 계좌 2의 잔액은 2000

---

문제 4. 아래의 SQL에 대한 설명으로 올바른 것은?
```sql
CREATE USER new_user IDENTIFIED BY password;

GRANT SELECT, INSERT, UPDATE ON schema_name.table_name TO new_user;

REVOKE UPDATE ON schema_name.table_name FROM new_user;
```
1. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
2. 기존 사용자에게 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
3. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 추가한다.
4. 새로운 사용자를 생성하고, 특정 스키마에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 스키마에 대한 UPDATE 권한만 취소한다.

**Comment by Psjepark** [comment, marked as answer]
> 문제 1.
> 
> 상품 테이블(PRODUCT)에는 상품 ID(id), 상품명(name), 가격(price)이 포함되어 있습니다. 상품 가격이 변경되면, 변경된 정보를 변경된 가격정보 테이블(UPDATE_PRODUCT)에 추가합니다. `UPDATE_PRODUCT`테이블은 상품 ID(update_id), 상품명(update_name), 변경된 가격(update_price)이 포함되어 있습니다. 변경된 가격 정보를 `PRODUCT`의 해당 상품 ID에 업데이트하고, 상품이 존재하지 않는 경우 새로운 상품을 추가하는 동기화 작업을 하려고 합니다. 상품 가격 변경 정보를 처리하는 MERGE 구문을 작성하세요.
> 
> > JHKIM.PRODUCT 및 JHKIM.UPDATE_PRODUCT 정보를 참고해도 됩니다...
> 
> 문제 2. 다음 보기 중 트랜잭션(Transaction) 관련 설명으로 올바르지 않은 것은?
> 
> 1. 트랜잭션은 데이터베이스에서 하나 이상의 작업을 그룹화하여 원자성(Atomicity), 일관성(Consistency), 격리성(Isolation), 지속성(Durability)을 보장한다.
> 2. 원자성(Atomicity)은 트랜잭션의 작업들이 모두 성공적으로 완료되거나, 아예 실행되지 않은 상태로 되돌려지는 것을 보장한다.
> 3. 격리성(Isolation)은 트랜잭션의 작업들이 다른 트랜잭션의 작업에 영향을 주지 않도록 보장한다.
> 4. 지속성(Durability)은 트랜잭션의 작업이 성공적으로 완료된 후, 해당 작업의 결과가 영구적으로 소실될 수 있다.
>  4

> 문제 3. 아래의 SQL 블록 실행 후, 각 계좌의 잔액은 어떻게 될까?
> 
> ```sql
> CREATE TABLE account (
>     id NUMBER PRIMARY KEY,
>     balance NUMBER
> );
> 
> INSERT INTO account (id, balance) VALUES (1, 1000);
> INSERT INTO account (id, balance) VALUES (2, 2000);
> 
> DECLARE
>     from_account_id NUMBER := 1;
>     to_account_id NUMBER := 2;
>     transfer_amount NUMBER := 500;
> BEGIN
>     UPDATE account SET balance = balance - transfer_amount WHERE id = from_account_id;
>     -- 에러발생!!
>     UPDATE account SET balance = balance + transfer_amount WHERE id = to_account_id;
>     ROLLBACK;
> END;
> ```
> 
> 1. 계좌 1의 잔액은 500, 계좌 2의 잔액은 2500
> 2. 계좌 1의 잔액은 1000, 계좌 2의 잔액은 2000
> 3. 계좌 1의 잔액은 0, 계좌 2의 잔액은 3000
> 4. 계좌 1의 잔액은 1500, 계좌 2의 잔액은 2000
>  
> 2 

> 문제 4. 아래의 SQL에 대한 설명으로 올바른 것은?
> 
> ```sql
> CREATE USER new_user IDENTIFIED BY password;
> 
> GRANT SELECT, INSERT, UPDATE ON schema_name.table_name TO new_user;
> 
> REVOKE UPDATE ON schema_name.table_name FROM new_user;
> ```
> 
> 1. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
> 2. 기존 사용자에게 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
> 3. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 추가한다.
> 4. 새로운 사용자를 생성하고, 특정 스키마에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 스키마에 대한 UPDATE 권한만 취소한다.

> 1

**Comment by white-gyu** [comment]
문제 1.  

상품 테이블(PRODUCT)에는 상품 ID(id), 상품명(name), 가격(price)이 포함되어 있습니다. 상품 가격이 변경되면, 변경된 정보를 변경된 가격정보 테이블(UPDATE_PRODUCT)에 추가합니다. `UPDATE_PRODUCT`테이블은 상품 ID(update_id), 상품명(update_name), 변경된 가격(update_price)이 포함되어 있습니다. 변경된 가격 정보를 `PRODUCT`의 해당 상품 ID에 업데이트하고, 상품이 존재하지 않는 경우 새로운 상품을 추가하는 동기화 작업을 하려고 합니다.  상품 가격 변경 정보를 처리하는 MERGE 구문을 작성하세요. 
> JHKIM.PRODUCT 및 JHKIM.UPDATE_PRODUCT 정보를 참고해도 됩니다...

```oracle
merge
    into jhkim.PRODUCT
    using jhkim.UPDATE_PRODUCT
        on (PRODUCT.ID = UPDATE_PRODUCT.UPDATE_ID)
    when matched then
        update
            set PRICE = UPDATE_PRICE
    when not matched then
        insert(PRODUCT.id, PRODUCT.name, PRODUCT.PRICE)
        values(UPDATE_ID, UPDATE_NAME, UPDATE_PRICE);
```

---

문제 2. 다음 보기 중 트랜잭션(Transaction) 관련 설명으로 올바르지 않은 것은?
1. 트랜잭션은 데이터베이스에서 하나 이상의 작업을 그룹화하여 원자성(Atomicity), 일관성(Consistency), 격리성(Isolation), 지속성(Durability)을 보장한다.
2. 원자성(Atomicity)은 트랜잭션의 작업들이 모두 성공적으로 완료되거나, 아예 실행되지 않은 상태로 되돌려지는 것을 보장한다.
3. 격리성(Isolation)은 트랜잭션의 작업들이 다른 트랜잭션의 작업에 영향을 주지 않도록 보장한다.
4. 지속성(Durability)은 트랜잭션의 작업이 성공적으로 완료된 후, 해당 작업의 결과가 영구적으로 소실될 수 있다.

> 4

---

문제 3. 아래의 SQL 블록 실행 후, 각 계좌의 잔액은 어떻게 될까?
```sql
CREATE TABLE account (
    id NUMBER PRIMARY KEY,
    balance NUMBER
);

INSERT INTO account (id, balance) VALUES (1, 1000);
INSERT INTO account (id, balance) VALUES (2, 2000);

DECLARE
    from_account_id NUMBER := 1;
    to_account_id NUMBER := 2;
    transfer_amount NUMBER := 500;
BEGIN
    UPDATE account SET balance = balance - transfer_amount WHERE id = from_account_id;
    -- 에러발생!!
    UPDATE account SET balance = balance + transfer_amount WHERE id = to_account_id;
    ROLLBACK;
END;
```

1. 계좌 1의 잔액은 500, 계좌 2의 잔액은 2500
2. 계좌 1의 잔액은 1000, 계좌 2의 잔액은 2000
3. 계좌 1의 잔액은 0, 계좌 2의 잔액은 3000
4. 계좌 1의 잔액은 1500, 계좌 2의 잔액은 2000

> 2

---

문제 4. 아래의 SQL에 대한 설명으로 올바른 것은?
```sql
CREATE USER new_user IDENTIFIED BY password;

GRANT SELECT, INSERT, UPDATE ON schema_name.table_name TO new_user;

REVOKE UPDATE ON schema_name.table_name FROM new_user;
```
1. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
2. 기존 사용자에게 특정 테이블에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 취소한다.
3. 새로운 사용자를 생성하고, 특정 테이블에 대한 SELECT, INSERT 권한을 부여한 다음, 해당 테이블에 대한 UPDATE 권한만 추가한다.
4. 새로운 사용자를 생성하고, 특정 스키마에 대한 SELECT, INSERT, UPDATE 권한을 부여한 다음, 해당 스키마에 대한 UPDATE 권한만 취소한다.

> 1
